brainiak/reconstruct/iem.py

```diff
@@ -0,0 +1,290 @@
+    in Visual Cortex. J. Neurosci. 33, 16275–16284 (2013).
+
+    [Bouwer2009] "1.Brouwer, G. J. & Heeger, D. J.
+    Decoding and Reconstructing Color from Responses in Human Visual
+    Cortex. J. Neurosci. 29, 13992–14003 (2009).
+"""
+
+# Authors: David Huberdeau (Yale University) &
+# Peter Kok (Yale University), 2018
+
+import logging
+import numpy as np
+from sklearn import linear_model
+from sklearn.base import BaseEstimator
+import math
+
+__all__ = [
+    "InvertedEncoding"
+]
+
+logger = logging.getLogger(__name__)
+
+
+class InvertedEncoding(BaseEstimator):
+    """Basis function-based reconstruction method
+
+    Inverted encoding models (alternatively known as forward
+    models) are used to reconstruct a feature, e.g. color of
+    a stimulus, from patterns across voxels in functional
+    data. The model uses n_channels number of idealized
+    basis functions and assumes that the transformation from
+    stimulus feature (e.g. color) to basis function is one-
+    to-one and invertible. The response of a voxel is
+    expressed as the weighted sum of basis functions.
+    In this implementation, basis functions were half-wave
+    sinusoids raised to an even power (in this case, 6).
+
+    To use this model, estimate the weights of the basis
+    functions separately for each voxel via linear
+    regression (using fit()). Then, as part of predict(),
+    compute channel outputs for new functional data (done
+    by _predict_channel_responses()), and associate those
+    responses to a feature type. Score() computes a measure
+    of the error of the prediction based on known ground
+    truth.
+
+    This implementation assumes a circular (or half-
+    circular) feature domain. Future implementations might
+    generalize the feature input space, and increase the
+    possible dimensionality.
+
+    Parameters
+    ----------
+    n_channels: int, default 5, number of channels
+        The number of channels, or basis functions, to be used
+        in the inverted encoding model
+
+    range_start: double, default 0, beginning of range of
+        independent variable (usually degrees)
+
+    range_stop: double, default 180, end of range of
+        independent variable (usually degrees)
+
+
+    Attributes
+    ----------
+    C_: [n_channels, channel density] NumPy 2D array
+        matrix defining channel values
+
+    C_D_: [n_channels, channel density] NumPy 2D array
+        matrix defining channel independent variable (usually
+        in degrees)
+
+    W_: sklearn.linear_model model containing weight matrix that
+        relates estimated channel responses to response amplitude
+        data
+    """
+    def __init__(self, n_channels=5, range_start=0, range_stop=180):
+
+        self.n_channels = n_channels  # default = 5
+        self.range_start = range_start  # in degrees, 0 - 360, def=0
+        self.range_stop = range_stop  # in degrees, 0 - 360, def=180
```

Comment:
I guess it would be nice to also do input validation for `range_start` and `range_stop`?

Reply:
Yes. I changed this, but also changed the comments to indicate that the values are more flexible. It now checks just that range_start is less than range_stop.